spec/endtoend/endtoend_spec.rb

```diff
@@ -0,0 +1,141 @@
+require 'open-uri'
+require 'openssl'
+require 'git_helper'
+require 'tsuru_helper'
+
+describe "TsuruEndToEnd" do
+  context "deploying an application" do
+    before(:all) do
+      @tsuru_home = Tempdir.new('tsuru-command')
+      @tsuru_command = TsuruCommandLine.new(
+        {'HOME' => @tsuru_home.path },
+        { :verbose => RSpec.configuration.verbose }
+      )
+
+      @tsuru_api_url = "https://#{RSpec.configuration.target_api_host}"
+      @tsuru_api_url_insecure = "http://#{RSpec.configuration.target_api_host}:8080"
+
+      @tsuru_command.target_add("ci", @tsuru_api_url)
+      @tsuru_command.target_add("ci-insecure", @tsuru_api_url_insecure)
+
+      @tsuru_user = RSpec.configuration.tsuru_user || raise("You must set 'TSURU_USER' env var")
+      @tsuru_pass = RSpec.configuration.tsuru_pass || raise("You must set 'TSURU_PASS' env var")
+
+      # Generate the ssh key and setup ssh
+      @ssh_id_rsa_path = File.join(@tsuru_home, '.ssh', 'id_rsa')
+      @ssh_id_rsa_pub_path = File.join(@tsuru_home, '.ssh', 'id_rsa.pub')
+      @ssh_config_file = File.join(@tsuru_home, '.ssh', 'config')
+      @ssh_wrapper_path = File.join(@tsuru_home, 'ssh-wrapper')
+      SshHelper.generate_key(@ssh_id_rsa_path)
+      SshHelper.write_config(
+        @ssh_config_file,
+        {
+        "StrictHostKeyChecking" => "no",
+        "IdentityFile" => @ssh_id_rsa_path,
+        }
+      )
+      SshHelper.write_ssh_wrapper(@ssh_wrapper_path, @ssh_config_file)
+
+      # Clone the sample app and setup Git
+      @sampleapp_name = 'sampleapp' + Time.now.to_i.to_s
+      @sampleapp_path = File.join(@tsuru_home, @sampleapp_name)
+
+      @git_command = GitCommandLine.new(@sampleapp_path,
+        {
+          'HOME' => @tsuru_home.path,
+          'GIT_SSH' => @ssh_wrapper_path
+        },
+        { :verbose => RSpec.configuration.verbose }
+      )
+      @git_command.clone("https://github.com/alphagov/flask-sqlalchemy-postgres-heroku-example.git")
+
+
+    end
+
+    after(:all) do
+      @tsuru_command.key_remove('rspec') # Remove previous state if needed
+      @tsuru_command.service_unbind('sampleapptestdb', @sampleapp_name)
```

Comment:
We use ` 'sampleapp' + Time.now.to_i.to_s` to workaround the issue in postgresapisql that it creates the role when binding but it does not delete it while unbind. The Role in postgres is somehow based on the name of the app. 

But you are right, if we randomise the db instance we can even run several instances of the test suite in parallel. 

I will do so.